File: frontend/src/api/laps.ts
import { apiClient } from './client';
import type { Lap } from '@/types';

export const lapsApi = {
  add: (booking_id: number, lap_number: number, lap_time: number) =>
    apiClient.post<Lap>('/laps/', { booking_id, lap_number, lap_time }).then(r => r.data),
  getByBooking: (booking_id: number) =>
    apiClient.get<Lap[]>('/laps/', { params: { booking_id } }).then(r => r.data),
};

File: backend/app/api/v1/endpoints/laps.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from typing import List

from app.db.session import get_db
from app.core.deps import get_current_user, get_current_admin
from app.modules.laps.service import LapService
from app.modules.laps.schemas import LapCreate, LapOut
from app.modules.users.models import User

router = APIRouter(prefix="/laps", tags=["laps"])


@router.post("/", response_model=LapOut, status_code=201)
def add_lap(
    body: LapCreate,
    db: Session = Depends(get_db),
    _admin: User = Depends(get_current_admin),   # admin-only
):
    """Record a lap. Admin-only. Only allowed during the session's active time window (MSK)."""
    return LapService(db).add_lap(body.booking_id, body.lap_number, body.lap_time)


@router.get("/", response_model=List[LapOut])
def get_laps(
    booking_id: int = Query(...),
    db: Session = Depends(get_db),
    _u: User = Depends(get_current_user),
):
    return LapService(db).get_laps_for_booking(booking_id)
